mailet/standard/src/main/java/org/apache/james/transport/mailets/StripAttachment.java

```diff
@@ -221,122 +228,157 @@ public void service(Mail mail) throws MailetException {
-                                    (ArrayList<String>) c);
-                        }
-                        c.add(filename);
-                    }
-                }
-                if (attributeName != null) {
-                    @SuppressWarnings("unchecked")
-                    Map<String, byte[]> m = (Map<String, byte[]>) mail.getAttribute(attributeName);
-                    if (m == null) {
-                        m = new LinkedHashMap<String, byte[]>();
-                        mail.setAttribute(attributeName, (LinkedHashMap<String, byte[]>) m);
-                    }
-                    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-                    OutputStream os = new BufferedOutputStream(
-                            byteArrayOutputStream);
-                    part.writeTo(os);
-                    m.put(fileName, byteArrayOutputStream.toByteArray());
-                }
-                if (removeAttachments.equals(REMOVE_MATCHED)) {
-                    ret = true;
-                }
-            }
-            if (!ret) {
-                ret = removeAttachments.equals(REMOVE_ALL);
+    private boolean shouldBeRemoved(BodyPart bodyPart, Mail mail) throws MessagingException, Exception {
+        String fileName = getFilename(bodyPart);
+        if (fileName == null) {
+            return false;
+        }
+
+        boolean shouldRemove = removeAttachments.equals(REMOVE_ALL);
+        if (fileNameMatches(fileName)) {
+            storeBodyPartAsFile(bodyPart, mail, fileName);
+            storeBodyPartAsMailAttribute(bodyPart, mail, fileName);
+            if (removeAttachments.equals(REMOVE_MATCHED)) {
+                shouldRemove = true;
             }
-            if (ret) {
-                @SuppressWarnings("unchecked")
-                Collection<String> c = (Collection<String>) mail
-                        .getAttribute(REMOVED_ATTACHMENTS_ATTRIBUTE_KEY);
-                if (c == null) {
-                    c = new ArrayList<String>();
-                    mail.setAttribute(REMOVED_ATTACHMENTS_ATTRIBUTE_KEY,
-                            (ArrayList<String>) c);
-                }
-                c.add(fileName);
+        }
+        storeFileNameAsAttribute(mail, fileName, shouldRemove);
+        return shouldRemove;
+    }
+
+    private void storeBodyPartAsFile(BodyPart bodyPart, Mail mail, String fileName) throws Exception {
+        if (directoryName != null) {
+            Optional<String> filename = saveAttachmentToFile(bodyPart, Optional.of(fileName));
+            if (filename.isPresent()) {
+                addFilenameToAttribute(mail, filename.get(), SAVED_ATTACHMENTS_ATTRIBUTE_KEY);
             }
         }
-        return ret;
+    }
+
+    private void addFilenameToAttribute(Mail mail, String filename, String attributeName) {
+        @SuppressWarnings("unchecked")
+        ArrayList<String> attributeValues = (ArrayList<String>) mail.getAttribute(attributeName);
```

Comment:
I don't think you need the precise implementation here, you can replace twice by List<String>